src/net/sourceforge/kolmafia/session/MallPriceManager.java

```diff
@@ -154,92 +154,81 @@ public static MallSearchRequest newMallSearchRequest(String category, String tie
+    Iterator<PurchaseRequest> i = search.iterator();
+    while (i.hasNext()) {
+      PurchaseRequest purchase = i.next();
+      if (purchase instanceof MallPurchaseRequest) {
+        MallPurchaseRequest mallPurchase = (MallPurchaseRequest) purchase;
+        if (shopId == mallPurchase.getShopId()) {
+          i.remove();
+          MallPriceManager.updateMallPrice(ItemPool.get(itemId), search);
+          return true;
+        }
       }
+    }
+    return false;
+  }
 
-      if (itemId != -1 && search.get(0).getItemId() != itemId) {
-        continue;
+  public static final void flushCache(final int itemId, final int shopId) {
+    // Remove shop from search results for a single item
+    if (itemId != -1) {
+      List<PurchaseRequest> search = MallPriceManager.mallSearches.get(itemId);
+      if (search != null && MallPriceManager.removeShopPurchaseRequest(itemId, shopId, search)) {
+        if (search.size() == 0) {
+          MallPriceManager.mallSearches.remove(itemId);
+        }
       }
+      return;
+    }
 
-      Iterator<PurchaseRequest> i2 = search.iterator();
-      while (i2.hasNext()) {
-        PurchaseRequest purchase = i2.next();
-        if (purchase instanceof MallPurchaseRequest) {
-          MallPurchaseRequest mallPurchase = (MallPurchaseRequest) purchase;
-          if (shopId == mallPurchase.getShopId()) {
-            i2.remove();
-            MallPriceManager.updateMallPrice(ItemPool.get(itemId), search);
-            if (itemId != -1) {
-              return;
-            }
-            break;
-          }
+    // Remove shop from search results for all items
+    Iterator<Entry<Integer, List<PurchaseRequest>>> i =
+        MallPriceManager.mallSearches.entrySet().iterator();
+    while (i.hasNext()) {
+      Entry<Integer, List<PurchaseRequest>> entry = i.next();
+      int key = entry.getKey();
+      List<PurchaseRequest> search = entry.getValue();
+      if (MallPriceManager.removeShopPurchaseRequest(key, shopId, search)) {
+        if (search.size() == 0) {
+          i.remove();
         }
       }
     }
   }
 
   public static final void flushCache(final int itemId) {
-    Iterator<List<PurchaseRequest>> i = MallPriceManager.mallSearches.values().iterator();
-    while (i.hasNext()) {
-      List<PurchaseRequest> search = i.next();
-      // Always remove empty searches
-      if (search == null || search.size() == 0) {
-        i.remove();
-        continue;
-      }
-      int id = search.get(0).getItemId();
-      if (itemId == id) {
-        i.remove();
-        MallPriceManager.updateMallPrice(ItemPool.get(itemId), search);
-        return;
-      }
-      break;
+    List<PurchaseRequest> search = MallPriceManager.mallSearches.get(itemId);
+    if (search != null) {
+      MallPriceManager.mallSearches.remove(itemId);
+      // *** Previous code had this. Really?
+      // MallPriceManager.updateMallPrice(ItemPool.get(itemId), search);
```

Comment:
I think removing the code entirely is better than commenting it out with a questioning comment.

If a bug occurs after this PR, someone can look at the code changes in the git history to potentially identify any issues.

Reply:
Done.